### PR TITLE
Try to make divisions behavior clearer

### LINKED ISSUE #8264: Improve documentation of `DataFrame.divisions`
Talking to a user recently, they were confused about how to pass `divisions=` in `set_index`. Specifically, they missed that the list had to have N+1 elements in it, and the N+1th element was the upper bound for the last partition.

The API docs currently link to https://docs.dask.org/en/latest/dataframe-design.html#partitions, which does describe this in words, but takes some close reading to get.

A simple example of setting `divisions` using a list of values, both in the `set_index` docs and in the dataframe design section, might be helpful.

### PR DESCRIPTION
**Pull request by jsignell**
- [x] Closes #8264
- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

I think this makes things clearer and it makes the example runnable without being huge.

### CODE REVIEW

File: dask/dataframe/core.py [hunk 1]
```diff
@@ -4328,18 +4328,27 @@ def set_index(
...
+        >>> divisions = pd.date_range(start="2021-01-01", end="2021-01-07", freq='1D')
+        ... divisions
+        DatetimeIndex(['2021-01-01', '2021-01-02', '2021-01-03', '2021-01-04',
+                    '2021-01-05', '2021-01-06', '2021-01-07'],
+                    dtype='datetime64[ns]', freq='D')
+
+        Note that ``len(divisons)`` is equal to ``npartitions + 1``. This is because ``divisions``
+        represents the upper and lower bounds of each partition.
+
+        >>> ddf2 = ddf.set_index("timestamp", sorted=True, divisions=divisions)
```
Comment (gjoseph92) [inline dask/dataframe/core.py:4351]: I wonder if the example would be easier to read if we used the `name` column and wrote the divisions by hand:
```python
        >>> divisions = ["Alice", "Frank", "Laura", "Quinn", "Ursula", "Zelda"]
        >>> ddf2 = ddf.set_index("name", divisions=divisions)
```
This doesn't illustrate `sorted=True`, but it does show a probably-more-common case of using a different unsorted column, and that writing divisions yourself is not that scary.

File: dask/dataframe/core.py [hunk 2]
```diff
@@ -4328,18 +4328,27 @@ def set_index(
...
         """
```
Comment (gjoseph92) [inline dask/dataframe/core.py:4352]: Not necessary, but might be nice to suggest a common interactive workflow:
```suggestion
        If you'll be running `set_index` on the same (or similar) datasets repeatedly,
        you could save time by letting Dask calculate good divisions once, then copy-pasting
        them to reuse. This is especially helpful running in a Jupyter notebook:

        >>> ddf2 = ddf.set_index("name")  # slow, calculates data distribution
        >>> ddf2.divisions
        ["Alice", "Laura", "Ursula", "Zelda"]
        # ^ Now copy-paste this and edit the line above to:
        # ddf2 = ddf.set_index("name", divisions=["Alice", "Laura", "Ursula", "Zelda"])
        """
```

File: dask/dataframe/core.py [hunk 3]
```diff
@@ -4310,7 +4310,7 @@ def set_index(
             See https://docs.dask.org/en/latest/dataframe-design.html#partitions
```
Comment (gjoseph92) [inline dask/dataframe/core.py:4310]:
```
        divisions: list, optional
            The "dividing lines" used to split the new index into partitions.
            For ``divisions=[0, 10, 50, 100]``, there would be three output partitions,
            where the new index contained [0, 10), [10, 50), and [50, 100), respectively.
            See https://docs.dask.org/en/latest/dataframe-design.html#partitions.
            If not given (default), good divisions are calculated by immediately computing
            the data and looking at the distribution of its values. For large datasets,
            this can be expensive.
            Note that if ``sorted=True``, specified divisions are assumed to match
            the existing partitions in the data; if this is untrue you should
            leave divisions empty and call ``repartition`` after ``set_index``.
```
Something like this might make the `divisions` parameter easier to understand for me, don't know if it would help anyone else?

**Review by gjoseph92 — COMMENTED** [review]
Thanks @jsignell! This does help clarify things quite a lot.

I have one other suggestion I couldn't comment on the lines for (noting that `npartitions` is ignored when `divisions` is given):
```python
        npartitions: int, None, or 'auto'
            The ideal number of output partitions. If None, use the same as
            the input. If 'auto' then decide by memory use.
            Only used when ``divisions`` is not given. If ``divisions`` is given,
            the number of output partitions will be ``len(divisions) - 1``.
```

File: dask/dataframe/core.py [hunk 4]
```diff
@@ -4328,18 +4328,27 @@ def set_index(
...
-        >>> divisions = pd.date_range('2000', '2010', freq='1D')
-        >>> df2 = df.set_index('timestamp', sorted=True, divisions=divisions)  # doctest: +SKIP
+        >>> divisions = pd.date_range(start="2021-01-01", end="2021-01-07", freq='1D')
+        ... divisions
+        DatetimeIndex(['2021-01-01', '2021-01-02', '2021-01-03', '2021-01-04',
+                    '2021-01-05', '2021-01-06', '2021-01-07'],
+                    dtype='datetime64[ns]', freq='D')
+
+        Note that ``len(divisons)`` is equal to ``npartitions + 1``. This is because ``divisions``
+        represents the upper and lower bounds of each partition.
```
[the remark this answers is not in the document]
Comment (jsignell) [reply dask/dataframe/core.py:4349]: I don't think it's possible to be too wordy. This seems helpful.

File: dask/dataframe/core.py (hunk 1 above)
Comment (jsignell) [reply dask/dataframe/core.py:4351]: Yeah I like that idea.
